client/driver/exec.go

```diff
@@ -41,12 +47,43 @@ func (d *ExecDriver) Fingerprint(cfg *config.Config, node *structs.Node) (bool,
 }
 
 func (d *ExecDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle, error) {
-	// Get the command
+	// Get the command to be ran, or if omitted, download an artifact for
+	// execution. Currently a supplied command takes precedence, and an artifact
+	// is only downloaded if no command is supplied
 	command, ok := task.Config["command"]
 	if !ok || command == "" {
 		return nil, fmt.Errorf("missing command for exec driver")
 	}
 
+	source, ok := task.Config["artifact_source"]
+	if ok && source != "" {
+		// Proceed to download an artifact to be executed.
+		// We use go-getter to support a variety of protocols, but need to change
+		// file permissions of the resulted download to be executable
+
+		// Create a location to download the artifact.
+		taskDir, ok := ctx.AllocDir.TaskDirs[d.DriverContext.taskName]
+		if !ok {
+			return nil, fmt.Errorf("[Err] driver.Exec: Could not find task directory for task: %v", d.DriverContext.taskName)
+		}
+		destDir := filepath.Join(taskDir, allocdir.TaskLocal)
+
+		artifactName := path.Base(source)
+		artifactFile := filepath.Join(destDir, artifactName)
+		if err := getter.GetFile(artifactFile, source); err != nil {
+			return nil, fmt.Errorf("[Err] driver.Exec: Error downloading source for Exec driver: %s", err)
+		}
+
+		// Add execution permissions to the newly downloaded artifact
+		cmd := exec.Command("chmod", "+x", artifactFile)
+		if err := cmd.Run(); err != nil {
+			log.Printf("[Err] driver.Exec: Error making artifact executable: %s", err)
+		}
+
+		// re-assign the command to be the local execution path
+		command = filepath.Join(allocdir.TaskLocal, command)
```

Comment:
You should test this by running it on your mac or windows but I think you have to update exec_universal.go to also set the working directory to the task dir otherwise I am not sure it will find the task. https://github.com/hashicorp/nomad/blob/master/client/executor/exec_universal.go#L38